src/coreclr/jit/lsraarm64.cpp

```diff
@@ -25,6 +25,262 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+//   consecutive `registersToCheck` are free or not.
+//
+// Arguments:
+//    regToAssign     - Register assigned to the first refposition.
+//    registersCount  - Number of registers to check.
+//    registerType    - Type of register.
+//
+//  Returns:
+//      True if all the consecutive registers starting from `regToAssign` were free. Even if one
+//      of them is busy, returns false.
+//
+bool LinearScan::areNextConsecutiveRegistersFree(regNumber regToAssign, int registersCount, var_types registerType)
+{
+    for (int i = 0; i < registersCount; i++)
+    {
+        if (isRegInUse(regToAssign, registerType))
+        {
+            return false;
+        }
+        regToAssign = regToAssign == REG_FP_LAST ? REG_FP_FIRST : REG_NEXT(regToAssign);
+    }
+
+    return true;
+}
+
+//------------------------------------------------------------------------
+// getFreeCandidates: Returns the mask of all the free candidates for given refPosition.
+//   If refPosition is the first RefPosition of a series of refpositions that needs
+//   consecutive registers, then returns only the mask such that it satisfies the need
+//   of having free consecutive registers.
+//
+// Arguments:
+//    candidates   - Register assigned to the first refposition.
+//    refPosition  - Number of registers to check.
+//
+//  Returns:
+//      Register mask of all the free registers
+//
+regMaskTP LinearScan::getFreeCandidates(regMaskTP candidates, RefPosition* refPosition)
+{
+    regMaskTP result = candidates & m_AvailableRegs;
+    if (!refPosition->isFirstRefPositionOfConsecutiveRegisters())
+    {
+        return result;
+    }
+
+    unsigned int registersNeeded   = refPosition->regCount;
+    regMaskTP    currAvailableRegs = result;
+    if (BitOperations::PopCount(currAvailableRegs) < registersNeeded)
+    {
+        // If number of free registers are less than what we need, no point in scanning
+        // for them.
+        return RBM_NONE;
+    }
+
+// At this point, for 'n' registers requirement, if Rm+1, Rm+2, Rm+3, ..., Rm+k are
+// available, create the mask only for Rm+1, Rm+2, ..., Rm+(k-n+1) to convey that it
+// is safe to assign any of those registers, but not beyond that.
+#define AppendConsecutiveMask(startIndex, endIndex, availableRegistersMask)                                            \
+    regMaskTP selectionStartMask = (1ULL << regAvailableStartIndex) - 1;                                               \
+    regMaskTP selectionEndMask   = (1ULL << (regAvailableEndIndex - registersNeeded + 1)) - 1;                         \
+    consecutiveResult |= availableRegistersMask & (selectionEndMask & ~selectionStartMask);                            \
+    overallResult |= availableRegistersMask;
+
+    regMaskTP overallResult          = RBM_NONE;
+    regMaskTP consecutiveResult      = RBM_NONE;
+    DWORD     regAvailableStartIndex = 0, regAvailableEndIndex = 0;
+    do
```

Comment:
This algorithm for finding sequences of registers doesn't consider "wrapping", like your allocation code does -- so that allocation code wrapping will never get run (presumably).

Reply:
Yes, that is a known problem and I think I have written a TODO somewhere. If not, will add a comment unless I can find an easy way to add it.

Reply:
Fixed this now.